File: example/1d-linear-convection/weno3/julia/01a/julia/boundary.jl
# julia/boundary.jl
# 目标：与 Python boundary.py 行为完全一致（1:1 同构）
# 前提：ist/ied 在 Julia 中按 1-based 设置（ist = nghosts + 1）

using NPZ  # 仅用于测试，实际模块可不依赖

# ---------------------- 辅助函数 ----------------------
"""
模拟 Python 的 getattr(obj, name, default)
"""
function getfield_safe(obj, name::Symbol, default)
    return isdefined(obj, name) ? getfield(obj, name) : default
end

# ---------------------- 抽象基类（Julia 风格：用函数接口） ----------------------
# Julia 无 abstract class，用文档+约定
# 所有子类型必须实现 apply!(bc, u)

# ---------------------- PeriodicBoundary ----------------------
mutable struct PeriodicBoundary
    cfd::Any
end

function apply!(self::PeriodicBoundary, u::Vector{Float64})
    nghosts = self.cfd.domain.nghosts
    ist = self.cfd.domain.ist   # 1-based, e.g., 3
    ied = self.cfd.domain.ied   # 1-based, e.g., 43

    # 左 ghost: u[ist - 1 - ig] = u[ied - 1 - ig]
    for ig in 0:(nghosts-1)
        u[ist - 1 - ig] = u[ied - 1 - ig]
    end
    # 右 ghost: u[ied + ig] = u[ist + ig]
    for ig in 0:(nghosts-1)
        u[ied + ig] = u[ist + ig]
    end
end

# ---------------------- DirichletBoundary ----------------------
mutable struct DirichletBoundary
    cfd::Any
end

function apply!(self::DirichletBoundary, u::Vector{Float64})
    nghosts = self.cfd.domain.nghosts
    ist = self.cfd.domain.ist
    ied = self.cfd.domain.ied

    left_value = getfield_safe(self.cfd.config, :left_boundary_value, 1.0)
    right_value = getfield_safe(self.cfd.config, :right_boundary_value, 2.0)

    # 左边界
    for ig in 0:(nghosts-1)
        u[ist - 1 - ig] = left_value
    end
    # 右边界
    for ig in 0:(nghosts-1)
        u[ied + ig] = right_value
    end

    # 调试信息
    if getfield_safe(self.cfd.config, :debug, false)
        println("  应用Dirichlet边界: 左值=$left_value, 右值=$right_value")
    end
end

# ---------------------- NeumannBoundary ----------------------
mutable struct NeumannBoundary
    cfd::Any
end

function apply!(self::NeumannBoundary, u::Vector{Float64})
    nghosts = self.cfd.domain.nghosts
    ist = self.cfd.domain.ist
    ied = self.cfd.domain.ied

    # 左边界零梯度：u[ist - 1 - ig] = u[ist + ig]
    for ig in 0:(nghosts-1)
        u[ist - 1 - ig] = u[ist + ig]
    end
    # 右边界零梯度：u[ied + ig] = u[ied - 1 - ig]  ← 注意是 ied - 1 - ig
    for ig in 0:(nghosts-1)
        u[ied + ig] = u[ied - 1 - ig]
    end

    # 调试信息
    if getfield_safe(self.cfd.config, :debug, false)
        println("  应用Neumann边界: 零梯度")
    end
end

File: example/1d-linear-convection/weno3/python-v1/02a/src/boundary.py
from abc import ABC, abstractmethod
from core.registry import ComponentRegistry, register_component

# ---------------------- 边界条件抽象基类（统一接口） ----------------------
class BoundaryCondition(ABC):
    """边界条件抽象基类：定义所有边界条件必须实现的接口"""
    def __init__(self, cfd):
        self.cfd = cfd
        self.domain = cfd.domain
        self.config = cfd.config  # 可从配置读取边界参数（如进口速度、固壁温度等）
    
    @abstractmethod
    def apply(self, u):
        """
        应用边界条件到解数组
        :param u: 包含ghost层的解数组（会直接修改该数组）
        :return: None
        """
        pass

# ---------------------- 具体边界条件实现（使用装饰器注册）--------------------

@register_component('boundary', 'periodic')
class PeriodicBoundary(BoundaryCondition):
    """周期边界条件（1D专用）"""
    def apply(self, u):
        nghosts = self.domain.nghosts
        ist = self.domain.ist
        ied = self.domain.ied
        
        # 左ghost层 = 右物理层
        for ig in range(nghosts):
            u[ist - 1 - ig] = u[ied - 1 - ig]
        
        # 右ghost层 = 左物理层
        for ig in range(nghosts):
            u[ied + ig] = u[ist + ig]
            
@register_component('boundary', 'dirichlet')
class DirichletBoundary(BoundaryCondition):
    """Dirichlet（固定值）边界条件（如进口固定速度、固壁零速度）"""
    def apply(self, u):
        nghosts = self.domain.nghosts
        ist = self.domain.ist
        ied = self.domain.ied
        
        # ✅ 修复：使用 getattr 而不是 .get() 方法
        # 旧代码: left_value = self.config.get("left_boundary_value", 1.0)
        # 新代码: 使用 getattr，它对于类和实例都适用
        left_value = getattr(self.config, "left_boundary_value", 1.0)
        
        # 左边界（进口）固定值
        for ig in range(nghosts):
            u[ist - 1 - ig] = left_value
        
        # ✅ 修复：同样使用 getattr
        right_value = getattr(self.config, "right_boundary_value", 2.0)
        
        # 右边界（出口）固定值
        for ig in range(nghosts):
            u[ied + ig] = right_value
        
        # 调试信息
        if hasattr(self.config, 'debug') and self.config.debug:
            print(f"  应用Dirichlet边界: 左值={left_value}, 右值={right_value}")
            
@register_component('boundary', 'neumann')
class NeumannBoundary(BoundaryCondition):
    """Neumann（零梯度）边界条件（如出口无梯度）"""
    def apply(self, u):
        nghosts = self.domain.nghosts
        ist = self.domain.ist
        ied = self.domain.ied
        
        # 左边界零梯度
        for ig in range(nghosts):
            u[ist - 1 - ig] = u[ist + ig]
        
        # 右边界零梯度
        for ig in range(nghosts):
            u[ied + ig] = u[ied - 1 - ig]
        
        # 调试信息
        if hasattr(self.config, 'debug') and self.config.debug:
            print(f"  应用Neumann边界: 零梯度")

class BoundaryConditionFactory:
    """边界条件工厂"""
    
    @staticmethod
    def create(cfd) -> 'BoundaryCondition':
        """
        创建边界条件实例
        
        Args:
            cfd: CFD对象
        
        Returns:
            边界条件实例
        """
        from core.registry import BaseFactory
        bc_type = cfd.config.boundary_type
        return BaseFactory.create_component('boundary', bc_type, cfd)
        
